File: hotel_mobile/lib/data/services/vnpay_service.dart
/**
 * VNPay Service - Xử lý thanh toán VNPay trong Flutter
 * 
 * Tự implement logic tạo payment URL (không dùng package bên thứ 3)
 * Dựa trên VNPay API documentation và backend implementation
 * 
 * Chức năng:
 * - Tạo payment URL trực tiếp từ Flutter (không cần backend API)
 * - Verify signature từ VNPay response
 * - Mở trình duyệt/WebView để thanh toán
 * - Xử lý kết quả thanh toán
 * - Query trạng thái giao dịch
 */

import 'dart:convert';
import 'package:dio/dio.dart';
import 'package:crypto/crypto.dart';
import 'package:intl/intl.dart';
import '../../core/constants/app_constants.dart';
import '../../core/config/payment_config.dart';
import '../../core/services/backend_auth_service.dart';
import '../../core/services/vnpay_signature_service.dart';

/// Model cho Bank (ngân hàng VNPay)
class VNPayBank {
  final String code;
  final String name;

  VNPayBank({
    required this.code,
    required this.name,
  });

  factory VNPayBank.fromJson(Map<String, dynamic> json) {
    return VNPayBank(
      code: json['code'] ?? '',
      name: json['name'] ?? '',
    );
  }
}

/// Service xử lý VNPay
/// Tự implement logic tạo payment URL (không dùng package)
class VNPayService {
  final Dio _dio;
  final BackendAuthService _authService = BackendAuthService();
  
  // VNPay Signature Service để verify response
  late final VNPaySignatureService _signatureService;

  VNPayService() : _dio = Dio(BaseOptions(
    baseUrl: AppConstants.baseUrl,
    connectTimeout: const Duration(seconds: 30),
    receiveTimeout: const Duration(seconds: 30),
    headers: {
      'Content-Type': 'application/json',
    },
  )) {
    // Khởi tạo VNPay Signature Service với hash secret từ PaymentConfig
    _signatureService = VNPaySignatureService(
      hashSecret: PaymentConfig.vnpayHashSecret,
    );
    
    // Add interceptor để thêm token vào header (cho các API khác)
    _dio.interceptors.add(InterceptorsWrapper(
      onRequest: (options, handler) async {
        final token = await _authService.getToken();
        if (token != null && token.isNotEmpty) {
          options.headers['Authorization'] = 'Bearer $token';
        }
        return handler.next(options);
      },
    ));
  }

  /// Format date theo format VNPay (yyyyMMddHHmmss)
  String _formatDate(DateTime date) {
    return DateFormat('yyyyMMddHHmmss').format(date);
  }

  /// Sanitize order info (loại bỏ ký tự đặc biệt)
  String _sanitizeOrderInfo(String orderInfo) {
    // VNPay chỉ chấp nhận: a-z, A-Z, 0-9, và các ký tự: - . _ ~
    return orderInfo.replaceAll(RegExp(r'[^a-zA-Z0-9\-._~]'), '');
  }

  /// Sort object theo thứ tự alphabet (giống backend)
  Map<String, String> _sortObject(Map<String, String> obj) {
    final sortedKeys = obj.keys.toList()..sort();
    final sorted = <String, String>{};
    for (final key in sortedKeys) {
      sorted[key] = obj[key]!;
    }
    return sorted;
  }

  /// Tạo query string từ params (không encode)
  String _createQueryString(Map<String, String> params) {
    return params.entries
        .map((e) => '${e.key}=${e.value}')
        .join('&');
  }

  /// Tạo URL thanh toán VNPay TRỰC TIẾP từ Flutter
  /// 
  /// Parameters:
  /// - [bookingId]: ID của booking cần thanh toán
  /// - [amount]: Số tiền (VND)
  /// - [orderInfo]: Mô tả đơn hàng
  /// - [bankCode]: Mã ngân hàng (optional, nếu muốn chọn ngân hàng cụ thể)
  /// - [bookingData]: Thông tin booking đầy đủ để lưu vào backend (optional)
  /// - [ipAddr]: IP address của client (optional, sẽ tự detect)
  /// 
  /// Returns: Map với paymentUrl và orderId
  Future<Map<String, dynamic>> createPaymentUrl({
    required int bookingId,
    required double amount,
    required String orderInfo,
    String? bankCode,
    Map<String, dynamic>? bookingData,
    String? ipAddr,
  }) async {
    try {
      print('📤 VNPay Service: Tạo payment URL trực tiếp từ Flutter');
      print('📋 VNPay Service: bookingId=$bookingId, amount=$amount');
      
      // Tạo order ID unique
      final now = DateTime.now();
      final orderId = 'BOOKING_${bookingId}_${now.millisecondsSinceEpoch}';
      
      // Lấy IP address (nếu không có, dùng default)
      final clientIp = ipAddr ?? '127.0.0.1';
      
      // Return URL - phải là public URL (không phải localhost)
      // Lấy từ backend API
      final returnUrl = await _getReturnUrl();
      
      if (returnUrl.contains('localhost') || returnUrl.contains('127.0.0.1')) {
        throw Exception('VNPay Sandbox không chấp nhận localhost làm Return URL. Vui lòng cấu hình Return URL công khai trong backend .env file.');
      }
      
      // Format dates
      final createDate = _formatDate(now);
      final expireDate = _formatDate(now.add(const Duration(minutes: 15)));
      
      // Tạo params theo đúng format VNPay
      final vnpParams = <String, String>{
        'vnp_Version': '2.1.0',
        'vnp_Command': 'pay',
        'vnp_TmnCode': PaymentConfig.vnpayTmnCode,
        'vnp_Amount': (amount * 100).toInt().toString(), // VNPay yêu cầu * 100
        'vnp_CurrCode': 'VND',
        'vnp_TxnRef': orderId,
        'vnp_OrderInfo': _sanitizeOrderInfo(orderInfo),
        'vnp_OrderType': 'billpayment',
        'vnp_Locale': 'vn',
        'vnp_ReturnUrl': returnUrl,
        'vnp_IpAddr': clientIp,
        'vnp_CreateDate': createDate,
        'vnp_ExpireDate': expireDate,
      };
      
      // Thêm bankCode nếu có
      if (bankCode != null && bankCode.trim().isNotEmpty) {
        vnpParams['vnp_BankCode'] = bankCode.trim();
      }
      
      // Sắp xếp params theo thứ tự alphabet (QUAN TRỌNG!)
      final sortedParams = _sortObject(vnpParams);
      
      // Tạo query string từ sorted params (không encode)
      final signData = _createQueryString(sortedParams);
      
      // Tạo HMAC SHA512 signature
      final key = utf8.encode(PaymentConfig.vnpayHashSecret);
      final bytes = utf8.encode(signData);
      final hmac = Hmac(sha512, key);
      final digest = hmac.convert(bytes);
      final signature = digest.toString();
      
      // Thêm signature vào params
      sortedParams['vnp_SecureHash'] = signature;
      
      // Tạo URL cuối cùng
      final queryString = _createQueryString(sortedParams);
      final baseUrl = PaymentConfig.useVnpaySandbox 
          ? PaymentConfig.vnpaySandboxUrl 
          : PaymentConfig.vnpayProductionUrl;
      final paymentUrl = '$baseUrl?$queryString';
      
      print('✅ VNPay Service: Payment URL đã được tạo thành công');
      print('📋 VNPay Service: Order ID: $orderId');
      print('📋 VNPay Service: Return URL: $returnUrl');
      print('📋 VNPay Service: Signature: ${signature.substring(0, 40)}...');
      
      // Lưu booking data vào backend (nếu có) - để backend xử lý sau khi payment success
      if (bookingData != null) {
        try {
          await _savePaymentInfo(bookingId, orderId, amount, bookingData);
        } catch (e) {
          print('⚠️ VNPay Service: Không thể lưu payment info vào backend: $e');
          // Không throw error, vẫn tiếp tục với payment
        }
      }
      
      return {
        'paymentUrl': paymentUrl,
        'orderId': orderId,
      };
    } catch (e) {
      print('❌ VNPay Service: Error tạo payment URL: $e');
      rethrow;
    }
  }
  
  /// Lấy Return URL từ backend hoặc config
  /// ⚠️ QUAN TRỌNG: Return URL phải là public URL (không phải localhost)
  Future<String> _getReturnUrl() async {
    try {
      // Thử lấy từ backend API để lấy Return URL đã được config
      // Backend sẽ trả về Return URL từ .env (đã là public URL)
      try {
        final response = await _dio.get('/api/v2/vnpay/config');
        
        // Nếu backend trả về error về localhost
        if (response.statusCode == 400 && response.data['error'] == 'INVALID_RETURN_URL') {
          final errorData = response.data;
          throw Exception(
            '${errorData['message']}\n\n'
            '${errorData['hint'] ?? ''}\n\n'
            'Ví dụ: ${errorData['example'] ?? ''}'
          );
        }
        
        if (response.data['success'] == true && response.data['data'] != null) {
          final returnUrl = response.data['data']['returnUrl'];
          if (returnUrl != null && !returnUrl.contains('localhost') && !returnUrl.contains('127.0.0.1')) {
            print('✅ VNPay Service: Lấy Return URL từ backend: $returnUrl');
            return returnUrl;
          } else {
            throw Exception('Backend trả về Return URL là localhost. Vui lòng cấu hình VNP_RETURN_URL trong file .env của backend với public URL và restart backend server.');
          }
        }
      } on DioException catch (e) {
        if (e.response != null && e.response!.statusCode == 400) {
          final errorData = e.response!.data;
          throw Exception(
            '${errorData['message'] ?? 'Lỗi cấu hình VNPay'}\n\n'
            '${errorData['hint'] ?? ''}\n\n'
            'Ví dụ: ${errorData['example'] ?? ''}'
          );
        }
        print('⚠️ VNPay Service: Không thể lấy Return URL từ backend: ${e.message}');
        rethrow;
      } catch (e) {
        print('⚠️ VNPay Service: Error lấy Return URL từ backend: $e');
        rethrow;
      }
      
      // Fallback: Dùng backend base URL (phải là public URL)
      final baseUrl = AppConstants.baseUrl;
      final returnUrl = '$baseUrl/api/payment/vnpay-return';
      
      // Kiểm tra nếu là localhost
      if (returnUrl.contains('localhost') || returnUrl.contains('127.0.0.1')) {
        throw Exception(
          'Backend URL đang là localhost.\n\n'
          'Vui lòng:\n'
          '1. Cấu hình VNP_RETURN_URL trong file .env của backend với public URL\n'
          '2. Hoặc cập nhật AppConstants.baseUrl trong Flutter với public URL\n'
          '3. Restart backend server sau khi cập nhật .env'
        );
      }
      
      return returnUrl;
    } catch (e) {
      print('❌ VNPay Service: Error lấy Return URL: $e');
      rethrow;
    }
  }
  
  /// Lưu thông tin payment vào backend (để backend xử lý IPN và tạo booking)
  Future<void> _savePaymentInfo(
    int bookingId,
    String orderId,
    double amount,
    Map<String, dynamic> bookingData,
  ) async {
    try {
      final token = await _authService.getToken();
      await _dio.post(
        '/api/v2/vnpay/save-payment-info',
        data: {
          'bookingId': bookingId,
          'orderId': orderId,
          'amount': amount,
          'bookingData': bookingData,
        },
        options: token != null ? Options(
          headers: {'Authorization': 'Bearer $token'},
        ) : null,
      );
      print('✅ VNPay Service: Đã lưu payment info vào backend');
    } catch (e) {
      print('⚠️ VNPay Service: Không thể lưu payment info: $e');
      // Không throw, chỉ log warning
    }
  }

  /// Lấy danh sách ngân hàng hỗ trợ VNPay
  /// 
  /// Returns: List các ngân hàng với code và name
  Future<List<VNPayBank>> getBankList() async {
    try {
      // Không cần token cho endpoint public này
      final response = await Dio(BaseOptions(
        baseUrl: AppConstants.baseUrl,
      )).get(PaymentConfig.vnpayGetBanksEndpoint);

      if (response.data['success'] == true) {
        final List<dynamic> banks = response.data['data'];
        return banks.map((bank) => VNPayBank.fromJson(bank)).toList();
      } else {
        throw Exception('Không thể lấy danh sách ngân hàng');
      }
    } on DioException catch (e) {
      print('❌ DioError getting VNPay bank list: ${e.message}');
      // Return default banks nếu API fail
      return _getDefaultBanks();
    } catch (e) {
      print('❌ Error getting VNPay bank list: $e');
      return _getDefaultBanks();
    }
  }

  /// Query trạng thái giao dịch VNPay
  /// 
  /// Parameters:
  /// - [orderId]: Mã đơn hàng từ VNPay
  /// - [transDate]: Ngày giao dịch (format: yyyyMMddHHmmss)
  /// 
  /// Returns: Map chứa thông tin giao dịch
  Future<Map<String, dynamic>> queryTransaction({
    required String orderId,
    required String transDate,
  }) async {
    try {
      final response = await _dio.post(
        PaymentConfig.vnpayQueryTransactionEndpoint,
        data: {
          'orderId': orderId,
          'transDate': transDate,
        },
      );

      if (response.data['success'] == true) {
        return response.data['data'];
      } else {
        throw Exception(response.data['message'] ?? 'Không thể truy vấn giao dịch');
      }
    } on DioException catch (e) {
      print('❌ DioError querying VNPay transaction: ${e.message}');
      if (e.response != null) {
        throw Exception(e.response?.data['message'] ?? 'Lỗi kết nối server');
      }
      throw Exception('Không thể kết nối đến server');
    } catch (e) {
      print('❌ Error querying VNPay transaction: $e');
      rethrow;
    }
  }

  /// Danh sách ngân hàng mặc định (fallback)
  List<VNPayBank> _getDefaultBanks() {
    return [
      VNPayBank(code: 'VNPAYQR', name: 'Cổng thanh toán VNPAYQR'),
      VNPayBank(code: 'VNBANK', name: 'Thanh toán qua ứng dụng hỗ trợ VNPAYQR'),
      VNPayBank(code: 'INTCARD', name: 'Thanh toán qua thẻ quốc tế'),
      VNPayBank(code: 'VIETCOMBANK', name: 'Vietcombank'),
      VNPayBank(code: 'VIETINBANK', name: 'VietinBank'),
      VNPayBank(code: 'BIDV', name: 'BIDV'),
      VNPayBank(code: 'AGRIBANK', name: 'Agribank'),
      VNPayBank(code: 'TECHCOMBANK', name: 'Techcombank'),
      VNPayBank(code: 'ACB', name: 'ACB'),
      VNPayBank(code: 'VPBANK', name: 'VPBank'),
      VNPayBank(code: 'MBBANK', name: 'MB Bank'),
      VNPayBank(code: 'SACOMBANK', name: 'Sacombank'),
    ];
  }

  /// Get signature service (để verify response)
  VNPaySignatureService get signatureService => _signatureService;
}

File: hotel-booking-backend/services/vnpayService.js
/**
 * VNPay Service - Theo đúng tài liệu VNPay chính thức
 * 
 * Tài liệu: https://sandbox.vnpayment.vn/apis/docs/thanh-toan-pay/pay.html
 * Version: 2.1.0
 * Algorithm: HMAC SHA512
 */

const crypto = require('crypto');
const querystring = require('querystring');
const vnpayConfig = require('../config/vnpay');

class VNPayService {
  /**
   * Tạo URL thanh toán VNPay
   * 
   * @param {Object} params
   * @param {string} params.orderId - Mã đơn hàng (unique)
   * @param {number} params.amount - Số tiền (VND)
   * @param {string} params.orderInfo - Mô tả đơn hàng
   * @param {string} params.ipAddr - IP address của khách hàng
   * @param {string} params.bankCode - Mã ngân hàng (optional)
   * @returns {string} URL thanh toán VNPay
   */
  createPaymentUrl(params) {
    const { orderId, amount, orderInfo, orderType = 'billpayment', ipAddr, bankCode = '' } = params;

    // Validate
    if (!orderId || !amount || !orderInfo) {
      throw new Error('Missing required fields: orderId, amount, orderInfo');
    }

    if (!vnpayConfig.vnp_TmnCode || !vnpayConfig.vnp_HashSecret) {
      throw new Error('VNPay config missing: vnp_TmnCode or vnp_HashSecret');
    }

    // Set timezone
    process.env.TZ = 'Asia/Ho_Chi_Minh';

    // Tạo thời gian
    const date = new Date();
    const createDate = this.formatDate(date);
    const expireDate = new Date(date.getTime() + vnpayConfig.vnp_ExpireTime * 60000);
    const expireDateStr = this.formatDate(expireDate);

    // Lấy IP address
    let clientIp = ipAddr || '127.0.0.1';
    if (clientIp.startsWith('::ffff:')) {
      clientIp = clientIp.substring(7);
    }
    if (clientIp === '::1') {
      clientIp = '127.0.0.1';
    }

    // Tạo các tham số VNPay
    const vnp_Params = {
      vnp_Version: vnpayConfig.vnp_Version,
      vnp_Command: vnpayConfig.vnp_Command,
      vnp_TmnCode: vnpayConfig.vnp_TmnCode,
      vnp_Amount: String(Math.round(amount * 100)), // VNPay yêu cầu * 100
      vnp_CurrCode: vnpayConfig.vnp_CurrCode,
      vnp_TxnRef: orderId,
      vnp_OrderInfo: this.sanitizeOrderInfo(orderInfo),
      vnp_OrderType: orderType,
      vnp_Locale: vnpayConfig.vnp_Locale,
      vnp_ReturnUrl: vnpayConfig.vnp_ReturnUrl,
      vnp_IpAddr: clientIp,
      vnp_CreateDate: createDate,
      vnp_ExpireDate: expireDateStr,
    };

    // Thêm bankCode nếu có
    if (bankCode && bankCode.trim() !== '') {
      vnp_Params['vnp_BankCode'] = bankCode.trim();
    }

    // Sắp xếp params theo alphabet (QUAN TRỌNG!)
    const sortedParams = this.sortObject(vnp_Params);

    // Tạo query string để tạo signature
    // Fix: Dùng URLSearchParams thay vì querystring (deprecated)
    const signData = new URLSearchParams(sortedParams).toString();

    // Tạo HMAC SHA512 signature
    const hmac = crypto.createHmac('sha512', vnpayConfig.vnp_HashSecret);
    const signed = hmac.update(signData, 'utf-8').digest('hex');

    // Thêm signature vào params
    sortedParams['vnp_SecureHash'] = signed;

    // Tạo payment URL với URLSearchParams
    const paymentQueryString = new URLSearchParams(sortedParams).toString();
    const vnpUrl = vnpayConfig.vnp_Url + '?' + paymentQueryString;

    console.log('✅ VNPay Payment URL created');
    console.log('   Order ID:', orderId);
    console.log('   Amount:', amount, 'VND');
    console.log('   Signature:', signed.substring(0, 40) + '...');

    return vnpUrl;
  }

  /**
   * Verify signature từ VNPay return
   * 
   * @param {Object} vnpParams - Params từ VNPay return
   * @returns {boolean} true nếu signature hợp lệ
   */
  verifyReturnUrl(vnpParams) {
    const secureHash = vnpParams.vnp_SecureHash;

    if (!secureHash) {
      console.error('❌ Missing vnp_SecureHash');
      return false;
    }

    // Tạo bản copy và xóa signature fields
    const paramsCopy = { ...vnpParams };
    delete paramsCopy.vnp_SecureHash;
    delete paramsCopy.vnp_SecureHashType;

    // Sắp xếp params
    const sortedParams = this.sortObject(paramsCopy);

    // Tạo query string để verify (giống createPaymentUrl)
    const signData = new URLSearchParams(sortedParams).toString();

    // Tạo signature để verify
    const hmac = crypto.createHmac('sha512', vnpayConfig.vnp_HashSecret);
    const signed = hmac.update(signData, 'utf-8').digest('hex');

    const isValid = secureHash === signed;

    if (!isValid) {
      console.error('❌ Signature mismatch');
      console.error('   Received:', secureHash);
      console.error('   Calculated:', signed);
      console.error('   SignData:', signData);
    } else {
      console.log('✅ Signature verified');
    }

    return isValid;
  }

  /**
   * Query transaction status
   * 
   * @param {Object} params
   * @param {string} params.orderId - Mã đơn hàng
   * @param {string} params.transDate - Ngày giao dịch (yyyyMMddHHmmss)
   * @returns {Promise<Object>} Kết quả query từ VNPay
   */
  async queryTransaction(params) {
    const { orderId, transDate } = params;
    process.env.TZ = 'Asia/Ho_Chi_Minh';
    const date = new Date();
    const createDate = this.formatDate(date);
    const requestId = this.generateRequestId();

    const data = {
      vnp_RequestId: requestId,
      vnp_Version: vnpayConfig.vnp_Version,
      vnp_Command: 'querydr',
      vnp_TmnCode: vnpayConfig.vnp_TmnCode,
      vnp_TxnRef: orderId,
      vnp_OrderInfo: `Query transaction ${orderId}`,
      vnp_TransactionDate: transDate,
      vnp_CreateDate: createDate,
      vnp_IpAddr: '127.0.0.1',
    };

    const sortedData = this.sortObject(data);
    const hashData = new URLSearchParams(sortedData).toString();
    
    const hmac = crypto.createHmac('sha512', vnpayConfig.vnp_HashSecret);
    const secureHash = hmac.update(hashData, 'utf-8').digest('hex');
    data.vnp_SecureHash = secureHash;

    try {
      const response = await fetch(vnpayConfig.vnp_Api, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(data),
      });
      return await response.json();
    } catch (error) {
      console.error('Error querying VNPay transaction:', error);
      throw error;
    }
  }

  /**
   * Sắp xếp object theo alphabet
   * 
   * @param {Object} obj - Object cần sắp xếp
   * @returns {Object} Object đã sắp xếp
   */
  sortObject(obj) {
    const sorted = {};
    const keys = Object.keys(obj).sort();
    
    for (const key of keys) {
      const value = obj[key];
      if (value !== null && value !== undefined && value !== '') {
        sorted[key] = String(value);
      }
    }
    
    return sorted;
  }

  /**
   * Sanitize orderInfo - Loại bỏ dấu tiếng Việt và ký tự đặc biệt
   * 
   * @param {string} orderInfo - Mô tả đơn hàng
   * @returns {string} Mô tả đã được sanitize
   */
  sanitizeOrderInfo(orderInfo) {
    if (!orderInfo) return '';
    
    let sanitized = String(orderInfo);
    
    // Chuyển tiếng Việt có dấu sang không dấu
    sanitized = this.removeVietnameseAccents(sanitized);
    
    // Loại bỏ ký tự đặc biệt, chỉ giữ lại: a-z, A-Z, 0-9, space, -, ., _
    sanitized = sanitized
      .replace(/[^a-zA-Z0-9\s\-._]/g, '')
      .replace(/\s+/g, ' ')
      .trim();
    
    // Giới hạn độ dài
    if (sanitized.length > 255) {
      sanitized = sanitized.substring(0, 255);
    }
    
    return sanitized;
  }

  /**
   * Chuyển tiếng Việt có dấu sang không dấu
   */
  removeVietnameseAccents(str) {
    if (!str) return '';
    
    const accentMap = {
      'à': 'a', 'á': 'a', 'ạ': 'a', 'ả': 'a', 'ã': 'a',
      'â': 'a', 'ầ': 'a', 'ấ': 'a', 'ậ': 'a', 'ẩ': 'a', 'ẫ': 'a',
      'ă': 'a', 'ằ': 'a', 'ắ': 'a', 'ặ': 'a', 'ẳ': 'a', 'ẵ': 'a',
      'è': 'e', 'é': 'e', 'ẹ': 'e', 'ẻ': 'e', 'ẽ': 'e',
      'ê': 'e', 'ề': 'e', 'ế': 'e', 'ệ': 'e', 'ể': 'e', 'ễ': 'e',
      'ì': 'i', 'í': 'i', 'ị': 'i', 'ỉ': 'i', 'ĩ': 'i',
      'ò': 'o', 'ó': 'o', 'ọ': 'o', 'ỏ': 'o', 'õ': 'o',
      'ô': 'o', 'ồ': 'o', 'ố': 'o', 'ộ': 'o', 'ổ': 'o', 'ỗ': 'o',
      'ơ': 'o', 'ờ': 'o', 'ớ': 'o', 'ợ': 'o', 'ở': 'o', 'ỡ': 'o',
      'ù': 'u', 'ú': 'u', 'ụ': 'u', 'ủ': 'u', 'ũ': 'u',
      'ư': 'u', 'ừ': 'u', 'ứ': 'u', 'ự': 'u', 'ử': 'u', 'ữ': 'u',
      'ỳ': 'y', 'ý': 'y', 'ỵ': 'y', 'ỷ': 'y', 'ỹ': 'y',
      'đ': 'd',
      'À': 'A', 'Á': 'A', 'Ạ': 'A', 'Ả': 'A', 'Ã': 'A',
      'Â': 'A', 'Ầ': 'A', 'Ấ': 'A', 'Ậ': 'A', 'Ẩ': 'A', 'Ẫ': 'A',
      'Ă': 'A', 'Ằ': 'A', 'Ắ': 'A', 'Ặ': 'A', 'Ẳ': 'A', 'Ẵ': 'A',
      'È': 'E', 'É': 'E', 'Ẹ': 'E', 'Ẻ': 'E', 'Ẽ': 'E',
      'Ê': 'E', 'Ề': 'E', 'Ế': 'E', 'Ệ': 'E', 'Ể': 'E', 'Ễ': 'E',
      'Ì': 'I', 'Í': 'I', 'Ị': 'I', 'Ỉ': 'I', 'Ĩ': 'I',
      'Ò': 'O', 'Ó': 'O', 'Ọ': 'O', 'Ỏ': 'O', 'Õ': 'O',
      'Ô': 'O', 'Ồ': 'O', 'Ố': 'O', 'Ộ': 'O', 'Ổ': 'O', 'Ỗ': 'O',
      'Ơ': 'O', 'Ờ': 'O', 'Ớ': 'O', 'Ợ': 'O', 'Ở': 'O', 'Ỡ': 'O',
      'Ù': 'U', 'Ú': 'U', 'Ụ': 'U', 'Ủ': 'U', 'Ũ': 'U',
      'Ư': 'U', 'Ừ': 'U', 'Ứ': 'U', 'Ự': 'U', 'Ử': 'U', 'Ữ': 'U',
      'Ỳ': 'Y', 'Ý': 'Y', 'Ỵ': 'Y', 'Ỷ': 'Y', 'Ỹ': 'Y',
      'Đ': 'D',
    };
    
    return str.split('').map(char => accentMap[char] || char).join('');
  }

  /**
   * Format date yyyyMMddHHmmss
   */
  formatDate(date) {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    const hour = String(date.getHours()).padStart(2, '0');
    const minute = String(date.getMinutes()).padStart(2, '0');
    const second = String(date.getSeconds()).padStart(2, '0');
    return `${year}${month}${day}${hour}${minute}${second}`;
  }

  /**
   * Generate request ID
   */
  generateRequestId() {
    return Date.now().toString() + Math.random().toString(36).substring(2, 9);
  }

  /**
   * Get response message từ response code
   */
  getResponseMessage(responseCode) {
    const messages = {
      '00': 'Giao dịch thành công',
      '07': 'Trừ tiền thành công. Giao dịch bị nghi ngờ.',
      '09': 'Thẻ/Tài khoản chưa đăng ký InternetBanking.',
      '10': 'Xác thực thông tin thẻ/tài khoản không đúng quá 3 lần',
      '11': 'Đã hết hạn chờ thanh toán',
      '12': 'Thẻ/Tài khoản bị khóa',
      '13': 'Nhập sai mật khẩu OTP',
      '24': 'Khách hàng hủy giao dịch',
      '51': 'Tài khoản không đủ số dư',
      '65': 'Vượt quá hạn mức giao dịch trong ngày',
      '75': 'Ngân hàng thanh toán đang bảo trì',
      '79': 'Nhập sai mật khẩu thanh toán quá số lần quy định',
      '99': 'Timeout/Lỗi VNPAY - VNPay không thể gọi về Return URL hoặc có lỗi hệ thống.',
    };
    return messages[responseCode] || 'Lỗi không xác định';
  }
}

module.exports = new VNPayService();

